packages/flutter/test/widgets/web_image_test.dart

```diff
@@ -0,0 +1,1308 @@
+        OneFrameWebImageStreamCompleter(
+          SynchronousFuture<bool>(false),
+          _imageCompleter.future,
+        );
+  }
+
+  final Completer<WebImageInfo> _imageCompleter = Completer<WebImageInfo>();
+  late WebImageStreamCompleter _streamCompleter;
+
+  bool get loadCalled => _loadCallCount > 0;
+  int get loadCallCount => _loadCallCount;
+  int _loadCallCount = 0;
+
+  @override
+  void resolveStreamForKey(
+      WebImageStream stream, String key, ImageErrorListener handleError) {
+    super.resolveStreamForKey(stream, key, handleError);
+  }
+
+  @override
+  WebImageStreamCompleter loadImage(String key) {
+    _loadCallCount += 1;
+    return _streamCompleter;
+  }
+
+  void complete(web_shim.HTMLImageElement image) {
+    _imageCompleter.complete(WebImageInfo(image));
+  }
+
+  void fail(Object exception, StackTrace? stackTrace) {
+    _imageCompleter.completeError(exception, stackTrace);
+  }
+}
+
+/// An [WebImageStreamCompleter] that gives access to the added listeners.
+///
+/// Such an access to listeners is hacky,
+/// because it breaks encapsulation by allowing to invoke listeners without
+/// taking care about lifecycle of the created images, that may result in not disposed images.
+///
+/// That's why some tests that use it
+/// are opted out from leak tracking.
+class _TestWebImageStreamCompleter extends WebImageStreamCompleter {
+  _TestWebImageStreamCompleter([this._currentImage]);
+
+  WebImageInfo? _currentImage;
+  final Set<WebImageStreamListener> listeners = <WebImageStreamListener>{};
+
+  @override
+  void addListener(WebImageStreamListener listener) {
+    listeners.add(listener);
+    if (_currentImage != null) {
+      listener.onImage(_currentImage!.clone(), true);
+    }
+  }
+
+  @override
+  void removeListener(WebImageStreamListener listener) {
+    listeners.remove(listener);
+  }
+
+  void setData({
+    WebImageInfo? imageInfo,
+  }) {
+    if (imageInfo != null) {
+      _currentImage = imageInfo;
+    }
+    final List<WebImageStreamListener> localListeners = listeners.toList();
+    for (final WebImageStreamListener listener in localListeners) {
+      if (imageInfo != null) {
+        listener.onImage(imageInfo.clone(), false);
+      }
+    }
+  }
+
+  void setError({
+    required Object exception,
+    StackTrace? stackTrace,
+  }) {
+    final List<WebImageStreamListener> localListeners = listeners.toList();
+    for (final WebImageStreamListener listener in localListeners) {
+      listener.onError?.call(exception, stackTrace);
+    }
+  }
+
+  void dispose() {
+    final List<WebImageStreamListener> listenersCopy = listeners.toList();
+    listenersCopy.forEach(removeListener);
+  }
+}
+
+class _FailingWebImageProvider extends WebImageProviderImpl {
+  _FailingWebImageProvider(super.src, {
+    this.failOnLoad = false,
+    required this.throws,
+    required this.image,
+  }) : assert(failOnLoad);
+
+  final bool failOnLoad;
+  final Object throws;
```

Comment:
```suggestion
  final Exception throws;
```

Per https://dart.dev/tools/linter-rules/only_throw_errors

Reply:
For normal code, I agree, but the tests that use this code check that the expected exception actually throws by checking the `toString()` of the thrown Object, which would make the tests dependent on the `toString` implementation of `Exception` if I made this change.

Reply:
Then I'm fine with leaving this as-is.

Reply:
Nevermind, using your other suggestion I was able to make this change. Done.